File: vignettes/14_graph_rag_structured/python/graph_rag_structured.py
"""
Structured GraphRAG — Python version

Maps tabular CSV data to an RDF knowledge graph using an OWL ontology,
then uses SPARQL retrieval + LLM generation for grounded answers.

Requirements:
    pip install rdflib openai
    ollama pull gemma3
"""

import csv
import re
from rdflib import Graph, Namespace, Literal, RDF, RDFS, URIRef, OWL
from openai import OpenAI

# ── Setup ─────────────────────────────────────────────────────────────────────

client = OpenAI(base_url="http://localhost:11434/v1", api_key="ollama")
MODEL = "gemma3:latest"

ONT = Namespace("http://example.org/ontology#")
RES = Namespace("http://example.org/resource#")

kg = Graph()
kg.bind("ont", ONT)
kg.bind("res", RES)


def to_iri(name: str) -> str:
    return re.sub(r"[^a-zA-Z0-9]+", "", name.strip())


# ── Step 1: Load Ontology ────────────────────────────────────────────────────

kg.parse("../data/jaguar_ontology.ttl", format="turtle")
print(f"Ontology loaded: {len(kg)} triples")

# ── Step 2: Load and Map CSV ─────────────────────────────────────────────────

with open("../data/jaguars.csv") as f:
    reader = csv.DictReader(f)
    records = list(reader)

print(f"Loaded {len(records)} jaguar records")

for record in records:
    jaguar_id = record["jaguar_id"]
    jaguar = RES[jaguar_id]

    # Core identity
    kg.add((jaguar, RDF.type, ONT.Jaguar))
    kg.add((jaguar, RDFS.label, Literal(record["name"])))
    kg.add((jaguar, ONT.scientificName, Literal("Panthera onca")))

    # Gender
    if record.get("gender"):
        kg.add((jaguar, ONT.hasGender, Literal(record["gender"])))

    # Locations (multi-valued)
    if record.get("location"):
        for loc in record["location"].split(";"):
            loc = loc.strip()
            loc_iri = RES[to_iri(loc)]
            kg.add((jaguar, ONT.occursIn, loc_iri))
            kg.add((loc_iri, RDF.type, ONT.Location))
            kg.add((loc_iri, RDFS.label, Literal(loc)))

    # Monitoring organizations (multi-valued)
    if record.get("monitoring_org"):
        for org in record["monitoring_org"].split(";"):
            org = org.strip()
            org_iri = RES[to_iri(org)]
            kg.add((jaguar, ONT.monitoredByOrg, org_iri))
            kg.add((org_iri, RDF.type, ONT.ConservationOrganization))
            kg.add((org_iri, RDFS.label, Literal(org)))

    # First sighted
    if record.get("first_sighted"):
        kg.add((jaguar, ONT.hasMonitoringStartDate, Literal(record["first_sighted"])))

    # Killed status
    if record.get("is_killed"):
        kg.add((jaguar, ONT.wasKilled, Literal(record["is_killed"].lower())))

    # Cause of death
    if record.get("cause_of_death"):
        kg.add((jaguar, ONT.causeOfDeath, Literal(record["cause_of_death"])))

    # Identification mark
    if record.get("identification_mark"):
        kg.add((jaguar, ONT.hasIdentificationMark, Literal(record["identification_mark"])))

    # Threats (multi-valued)
    if record.get("threats"):
        for threat in record["threats"].split(";"):
            threat = threat.strip()
            threat_iri = RES[to_iri(threat)]
            kg.add((jaguar, ONT.facesThreat, threat_iri))
            kg.add((threat_iri, RDF.type, ONT.Threat))
            kg.add((threat_iri, RDFS.label, Literal(threat)))

    # Monitoring techniques (multi-valued)
    if record.get("monitoring_technique"):
        for tech in record["monitoring_technique"].split(";"):
            tech = tech.strip()
            tech_iri = RES[to_iri(tech)]
            kg.add((jaguar, ONT.monitoredByTechnique, tech_iri))
            kg.add((tech_iri, RDF.type, ONT.MonitoringTechnique))
            kg.add((tech_iri, RDFS.label, Literal(tech)))

    # Status notes
    if record.get("status_notes"):
        kg.add((jaguar, RDFS.comment, Literal(record["status_notes"])))

print(f"Knowledge graph: {len(kg)} triples")
print()
print("Turtle serialization (first 2000 chars):")
print(kg.serialize(format="turtle")[:2000])

# ── Step 3: SPARQL Queries ───────────────────────────────────────────────────

print("\n=== Jaguars and their locations ===")
results = kg.query("""
PREFIX ont: <http://example.org/ontology#>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
SELECT ?name ?location WHERE {
    ?j a ont:Jaguar ; rdfs:label ?name ; ont:occursIn ?loc .
    ?loc rdfs:label ?location .
} ORDER BY ?name
""")
for row in results:
    print(f"  {row.name} → {row.location}")

print("\n=== Jaguars that were killed ===")
results = kg.query("""
PREFIX ont: <http://example.org/ontology#>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
SELECT ?name ?cause WHERE {
    ?j a ont:Jaguar ; rdfs:label ?name ; ont:wasKilled "true" .
    OPTIONAL { ?j ont:causeOfDeath ?cause }
}
""")
for row in results:
    cause = f" — {row.cause}" if row.cause else ""
    print(f"  {row.name}{cause}")


# ── Step 4: Retrieval Functions ──────────────────────────────────────────────

def query_jaguar(name: str) -> str:
    results = list(kg.query(f"""
    PREFIX ont: <http://example.org/ontology#>
    PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
    SELECT ?pred ?obj WHERE {{
        ?j a ont:Jaguar ; rdfs:label "{name}" ; ?pred ?obj .
    }}
    """))
    if not results:
        return f"No jaguar found with name '{name}'."
    lines = []
    for row in results:
        pred = str(row.pred).split("#")[-1].split("/")[-1]
        if pred == "type":
            continue
        lines.append(f"{pred}: {row.obj}")
    return f"Facts about {name}:\n" + "\n".join(lines)


def run_sparql_query(query: str) -> str:
    try:
        results = list(kg.query(query))
        if not results:
            return "Query returned no results."
        lines = []
        for row in results:
            parts = [f"{str(val)}" for val in row]
            lines.append(", ".join(parts))
        return f"Results ({len(results)} rows):\n" + "\n".join(lines)
    except Exception as e:
        return f"SPARQL error: {e}"


# ── Step 5: Structured GraphRAG Pipeline ─────────────────────────────────────

def ask_structured_rag(question: str, jaguars=None, sparql_query_str="") -> str:
    context_parts = []
    for name in (jaguars or []):
        context_parts.append(query_jaguar(name))
    if sparql_query_str:
        context_parts.append(run_sparql_query(sparql_query_str))
    context = "\n\n".join(context_parts)

    response = client.chat.completions.create(
        model=MODEL,
        messages=[
            {
                "role": "system",
                "content": f"""You are a wildlife conservation expert specializing in jaguars.
Answer the question using ONLY the knowledge graph facts provided below.
Do not use your own knowledge. If the facts don't contain enough information, say so.

Knowledge Graph Facts:
{context}""",
            },
            {"role": "user", "content": question},
        ],
        temperature=0.1,
    )
    return response.choices[0].message.content


# ── Step 6: Ask Questions ────────────────────────────────────────────────────

questions = [
    ("Tell me about El Jefe.", ["El Jefe"], ""),
    (
        "Which jaguars have been killed and what were the causes?",
        ["Macho B", "Yo'oko"],
        """PREFIX ont: <http://example.org/ontology#>
        PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
        SELECT ?name ?cause WHERE {
            ?j a ont:Jaguar ; rdfs:label ?name ; ont:wasKilled "true" .
            OPTIONAL { ?j ont:causeOfDeath ?cause }
        }""",
    ),
    (
        "What monitoring techniques are used to track jaguars?",
        [],
        """PREFIX ont: <http://example.org/ontology#>
        PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
        SELECT ?name ?technique WHERE {
            ?j a ont:Jaguar ; rdfs:label ?name ;
               ont:monitoredByTechnique ?t .
            ?t rdfs:label ?technique .
        }""",
    ),
]

for q, jags, sparql in questions:
    print("=" * 60)
    print(f"Q: {q}")
    print("-" * 60)
    answer = ask_structured_rag(q, jaguars=jags, sparql_query_str=sparql)
    print(f"A: {answer}")
    print()

File: vignettes/14_graph_rag_structured/csharp/GraphRagStructured.cs
// Structured GraphRAG — C# version
//
// Maps tabular CSV data to an RDF knowledge graph using an OWL ontology,
// then uses SPARQL retrieval + LLM generation for grounded answers.
//
// Requirements:
//   dotnet add package dotNetRDF
//   dotnet add package Microsoft.Extensions.AI
//   dotnet add package Microsoft.Extensions.AI.Ollama
//   ollama pull gemma3

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using VDS.RDF;
using VDS.RDF.Parsing;
using VDS.RDF.Query;
using VDS.RDF.Writing;
using Microsoft.Extensions.AI;
using OllamaSharp;

// ── Setup ────────────────────────────────────────────────────────────────────

IChatClient chatClient = new OllamaApiClient(new Uri("http://localhost:11434"), "gemma3:latest");

var store = new VDS.RDF.TripleStore();
var graph = new VDS.RDF.Graph();
store.Add(graph);

string ontNs = "http://example.org/ontology#";
string resNs = "http://example.org/resource#";
graph.NamespaceMap.AddNamespace("ont", new Uri(ontNs));
graph.NamespaceMap.AddNamespace("res", new Uri(resNs));

string ToIri(string name) => Regex.Replace(name.Trim(), @"[^a-zA-Z0-9]+", "");

// ── Step 1: Load Ontology ────────────────────────────────────────────────────

string ontologyPath = Path.Combine("..", "data", "jaguar_ontology.ttl");
var ttlParser = new TurtleParser();
ttlParser.Load(graph, ontologyPath);
Console.WriteLine($"Ontology loaded: {graph.Triples.Count} triples");

// ── Step 2: Load and Map CSV ─────────────────────────────────────────────────

string csvPath = Path.Combine("..", "data", "jaguars.csv");
var lines = File.ReadAllLines(csvPath);
var header = lines[0].Split(',');
var records = new List<Dictionary<string, string>>();

foreach (var line in lines.Skip(1))
{
    var fields = line.Split(',');
    var record = new Dictionary<string, string>();
    for (int i = 0; i < header.Length && i < fields.Length; i++)
        record[header[i].Trim()] = fields[i].Trim();
    records.Add(record);
}

Console.WriteLine($"Loaded {records.Count} jaguar records");

var rdf = graph.CreateUriNode(new Uri("http://www.w3.org/1999/02/22-rdf-syntax-ns#type"));
var rdfsLabel = graph.CreateUriNode(new Uri("http://www.w3.org/2000/01/rdf-schema#label"));
var rdfsComment = graph.CreateUriNode(new Uri("http://www.w3.org/2000/01/rdf-schema#comment"));

foreach (var record in records)
{
    string id = record.GetValueOrDefault("jaguar_id", "");
    if (string.IsNullOrEmpty(id)) continue;

    var jaguar = graph.CreateUriNode(new Uri(resNs + id));
    var jaguarType = graph.CreateUriNode(new Uri(ontNs + "Jaguar"));

    graph.Assert(new Triple(jaguar, rdf, jaguarType));
    graph.Assert(new Triple(jaguar, rdfsLabel,
        graph.CreateLiteralNode(record.GetValueOrDefault("name", ""))));
    graph.Assert(new Triple(jaguar, graph.CreateUriNode(new Uri(ontNs + "scientificName")),
        graph.CreateLiteralNode("Panthera onca")));

    // Gender
    string gender = record.GetValueOrDefault("gender", "");
    if (!string.IsNullOrEmpty(gender))
        graph.Assert(new Triple(jaguar, graph.CreateUriNode(new Uri(ontNs + "hasGender")),
            graph.CreateLiteralNode(gender)));

    // Locations (multi-valued)
    string location = record.GetValueOrDefault("location", "");
    if (!string.IsNullOrEmpty(location))
    {
        foreach (var loc in location.Split(';'))
        {
            string locTrimmed = loc.Trim();
            var locIri = graph.CreateUriNode(new Uri(resNs + ToIri(locTrimmed)));
            graph.Assert(new Triple(jaguar, graph.CreateUriNode(new Uri(ontNs + "occursIn")), locIri));
            graph.Assert(new Triple(locIri, rdf, graph.CreateUriNode(new Uri(ontNs + "Location"))));
            graph.Assert(new Triple(locIri, rdfsLabel, graph.CreateLiteralNode(locTrimmed)));
        }
    }

    // Monitoring organizations (multi-valued)
    string monitoringOrg = record.GetValueOrDefault("monitoring_org", "");
    if (!string.IsNullOrEmpty(monitoringOrg))
    {
        foreach (var org in monitoringOrg.Split(';'))
        {
            string orgTrimmed = org.Trim();
            var orgIri = graph.CreateUriNode(new Uri(resNs + ToIri(orgTrimmed)));
            graph.Assert(new Triple(jaguar, graph.CreateUriNode(new Uri(ontNs + "monitoredByOrg")), orgIri));
            graph.Assert(new Triple(orgIri, rdf, graph.CreateUriNode(new Uri(ontNs + "ConservationOrganization"))));
            graph.Assert(new Triple(orgIri, rdfsLabel, graph.CreateLiteralNode(orgTrimmed)));
        }
    }

    // First sighted date
    string firstSighted = record.GetValueOrDefault("first_sighted", "");
    if (!string.IsNullOrEmpty(firstSighted))
        graph.Assert(new Triple(jaguar, graph.CreateUriNode(new Uri(ontNs + "hasMonitoringStartDate")),
            graph.CreateLiteralNode(firstSighted)));

    // Killed status
    string isKilled = record.GetValueOrDefault("is_killed", "");
    if (!string.IsNullOrEmpty(isKilled))
        graph.Assert(new Triple(jaguar, graph.CreateUriNode(new Uri(ontNs + "wasKilled")),
            graph.CreateLiteralNode(isKilled.ToLower() == "true" ? "true" : "false")));

    // Cause of death
    string causeOfDeath = record.GetValueOrDefault("cause_of_death", "");
    if (!string.IsNullOrEmpty(causeOfDeath))
        graph.Assert(new Triple(jaguar, graph.CreateUriNode(new Uri(ontNs + "causeOfDeath")),
            graph.CreateLiteralNode(causeOfDeath)));

    // Identification mark
    string idMark = record.GetValueOrDefault("identification_mark", "");
    if (!string.IsNullOrEmpty(idMark))
        graph.Assert(new Triple(jaguar, graph.CreateUriNode(new Uri(ontNs + "hasIdentificationMark")),
            graph.CreateLiteralNode(idMark)));

    // Threats (multi-valued)
    string threats = record.GetValueOrDefault("threats", "");
    if (!string.IsNullOrEmpty(threats))
    {
        foreach (var threat in threats.Split(';'))
        {
            string t = threat.Trim();
            var threatIri = graph.CreateUriNode(new Uri(resNs + ToIri(t)));
            graph.Assert(new Triple(jaguar, graph.CreateUriNode(new Uri(ontNs + "facesThreat")), threatIri));
            graph.Assert(new Triple(threatIri, rdf, graph.CreateUriNode(new Uri(ontNs + "Threat"))));
            graph.Assert(new Triple(threatIri, rdfsLabel, graph.CreateLiteralNode(t)));
        }
    }

    // Monitoring techniques (multi-valued)
    string monTech = record.GetValueOrDefault("monitoring_technique", "");
    if (!string.IsNullOrEmpty(monTech))
    {
        foreach (var tech in monTech.Split(';'))
        {
            string techTrimmed = tech.Trim();
            var techIri = graph.CreateUriNode(new Uri(resNs + ToIri(techTrimmed)));
            graph.Assert(new Triple(jaguar, graph.CreateUriNode(new Uri(ontNs + "monitoredByTechnique")), techIri));
            graph.Assert(new Triple(techIri, rdf, graph.CreateUriNode(new Uri(ontNs + "MonitoringTechnique"))));
            graph.Assert(new Triple(techIri, rdfsLabel, graph.CreateLiteralNode(techTrimmed)));
        }
    }

    // Status notes
    string statusNotes = record.GetValueOrDefault("status_notes", "");
    if (!string.IsNullOrEmpty(statusNotes))
        graph.Assert(new Triple(jaguar, rdfsComment,
            graph.CreateLiteralNode(statusNotes)));
}

Console.WriteLine($"Knowledge graph: {graph.Triples.Count} triples");

// ── Step 3: SPARQL Queries ───────────────────────────────────────────────────

string RunSparql(string sparql)
{
    var processor = new VDS.RDF.Query.LeviathanQueryProcessor(store);
    var parser = new SparqlQueryParser();
    var query = parser.ParseFromString(sparql);
    var resultSet = processor.ProcessQuery(query) as SparqlResultSet;
    if (resultSet == null || resultSet.Count == 0)
        return "No results.";

    var sb = new StringBuilder();
    foreach (var result in resultSet)
    {
        var parts = result.Variables
            .Where(v => result.HasValue(v) && result[v] != null)
            .Select(v =>
            {
                var node = result[v];
                string val = node is ILiteralNode lit ? lit.Value
                    : node?.ToString()?.Split('/').Last() ?? "";
                return $"{v}={val}";
            });
        sb.AppendLine(string.Join(", ", parts));
    }
    return sb.ToString().TrimEnd();
}

Console.WriteLine("\n=== Jaguars and their locations ===");
Console.WriteLine(RunSparql(@"
PREFIX ont: <http://example.org/ontology#>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
SELECT ?name ?location WHERE {
    ?j a ont:Jaguar ; rdfs:label ?name ; ont:occursIn ?loc .
    ?loc rdfs:label ?location .
} ORDER BY ?name"));

// ── Step 4: Structured GraphRAG ──────────────────────────────────────────────

async Task<string> AskStructuredRag(string question, string context)
{
    var messages = new List<ChatMessage>
    {
        new(ChatRole.System, $@"You are a wildlife conservation expert specializing in jaguars.
Answer the question using ONLY the knowledge graph facts provided below.
Do not use your own knowledge. If the facts don't contain enough information, say so.

Knowledge Graph Facts:
{context}"),
        new(ChatRole.User, question),
    };

    var response = await chatClient.GetResponseAsync(messages,
        new ChatOptions { Temperature = 0.1f });
    return response.Text;
}

// ── Step 5: Ask Questions ────────────────────────────────────────────────────

string elJefeFacts = RunSparql(@"
PREFIX ont: <http://example.org/ontology#>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
SELECT ?pred ?obj WHERE {
    ?j a ont:Jaguar ; rdfs:label ""El Jefe"" ; ?pred ?obj .
}");

Console.WriteLine("\n=== Tell me about El Jefe ===");
Console.WriteLine(await AskStructuredRag("Tell me about El Jefe.", elJefeFacts));

string killedFacts = RunSparql(@"
PREFIX ont: <http://example.org/ontology#>
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
SELECT ?name ?cause WHERE {
    ?j a ont:Jaguar ; rdfs:label ?name ; ont:wasKilled ""true"" .
    OPTIONAL { ?j ont:causeOfDeath ?cause }
}");

Console.WriteLine("\n=== Which jaguars were killed? ===");
Console.WriteLine(await AskStructuredRag("Which jaguars have been killed and what were the causes?", killedFacts));
